src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf8Utility.Validation.cs

```diff
@@ -153,19 +164,17 @@ internal static unsafe partial class Utf8Utility
 
                         continue; // need to perform a bounds check because we might be running out of data
 
-                    Sse2LoopTerminatedEarlyDueToNonAsciiData:
-
-                        Debug.Assert(BitConverter.IsLittleEndian);
-                        Debug.Assert(Sse2.IsSupported);
+                    LoopTerminatedEarlyDueToNonAsciiData:
+                        // x86 can only be little endian, while ARM can be big or little endian
+                        // so if we reached this label we need to check both combinations are supported
+                        Debug.Assert((AdvSimd.Arm64.IsSupported && BitConverter.IsLittleEndian) || Sse2.IsSupported);
 
                         // The 'mask' value will have a 0 bit for each ASCII byte we saw and a 1 bit
                         // for each non-ASCII byte we saw. We can count the number of ASCII bytes,
                         // bump our input counter by that amount, and resume processing from the
                         // "the first byte is no longer ASCII" portion of the main loop.
 
-                        Debug.Assert(mask != 0);
-
-                        pInputBuffer += BitOperations.TrailingZeroCount(mask);
+                        pInputBuffer += trailingZeroCount;
```

Comment:
>trailingZeroCount [](start = 40, length = 17)

Add an assert here - `Debug.Assert(trailingZeroCount < 16)` so we don't accidently increase the buffer by more than that amount.

Reply:
If I add it there, it would affect both the AdvSimd and the Sse2 cases. This check should just be for the AdvSimd case, correct? In which case, I should add it to line 136 instead.